walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -55,4 +53,42 @@
     }
     return validTimes;
   }
+
+  public Collection<TimeRange> findOverlapTimes(Collection<TimeRange> mandatoryTimes, Collection<TimeRange> optionalTimes, long duration) {
+    Collection<TimeRange> validTimes = new ArrayList<>();
+    for(TimeRange optional : optionalTimes) {
+      for(TimeRange mandatory : mandatoryTimes) {
+        if(mandatory.overlaps(optional)) {
+          int start = Math.max(optional.start(), mandatory.start());
+          int end = Math.min(optional.end(), mandatory.end());
+          if(end == TimeRange.END_OF_DAY) {
+            end += 1;
+          }
+          TimeRange time = TimeRange.fromStartEnd(start, end, false);
+          if(time.duration() >= duration) {
+            validTimes.add(time);
+          }
+        }
```

Comment:
Good point on 

> both require some knowledge of how TimeRange works

And you might think why does FindMeetingQuery care how TimeRanges work, they just want a TimeRange and thats its.

Because of that, this block about getting the overlapping TimeRange of two TimeRange's seems like a good candidate for a public static method in the TimeRange class. Then you call it in this for loop like:

```
TimeRange overlap = TimeRange.getOverlap(optional, mandatory);
if (overlap != null && overlap.duration() >= duration) {
  validTimes.add(overlap);
}
```